### ISSUE TITLE
Creation modal for "New Page"

### ISSUE THREAD

**Issue by mtias** [WordPress/gutenberg#48456]
Let's explore combining the current "zoomed.out" experiment for patterns with the new frame mode when it comes to building new pages and templates based on sections. The idea behind zoomed.out was to allow operating at a higher level than individual blocks, which is now better accomplished in the frame mode.

### Iteration 1

_Can be triggered from nav menu, buttons, command center, etc._

- [x] Modal with page title, slug, and parent
- [x] Once in the editor, patterns modal shows

Mockups:

<img width="740" alt="Add modal" src="https://user-images.githubusercontent.com/1204802/236853621-aa85617a-34da-4f2f-87d6-b16e145da8d3.png">

<img width="1240" alt="near term" src="https://user-images.githubusercontent.com/1204802/237036077-74f7c6cb-d13f-4eec-ba9e-cff30951026c.png">

<img width="6784" alt="Adding pages flow" src="https://user-images.githubusercontent.com/1204802/236853799-fb4560ca-6f4d-43b8-a833-f24a22117d57.png">

See also [this comment for some more explorations on the slug editing](https://github.com/WordPress/gutenberg/issues/48456#issuecomment-1539662263), there is a simpler option if the above is too complex.

### Iteration 2: See #50739

[Underway](https://github.com/WordPress/gutenberg/issues/48456#issuecomment-1536460934).

**Comment by jameskoster** [comment]
Edit: The following concept speaks to "Iteration 2" as detailed in the OP.

Here's a potential design for the add page flow. It doesn't capture the working-with-sections element yet, but does install the basics.

https://user-images.githubusercontent.com/846565/236507894-337b60ee-533b-4480-9162-fe5112f926df.mp4

* The flow originates in the Pages panel (as proposed in https://github.com/WordPress/gutenberg/issues/44461#issuecomment-1536430171), where clicking the + reveals the New Page panel.
  * Worth noting that there could be a command in the Command Center that would send a user to this panel too
* The New Page panel contains a two fields:
  * Input for page title
  * Template selection modal.
* Once the page is created, any patterns associated with the Post Content block are presented in a modal (this is [existing functionality](https://github.com/WordPress/gutenberg/pull/40034)).

One technical detail to discuss here (in addition to any design feedback) is the status applied to these new pages on creation. They should probably be drafts, but there's not currently a way to _publish_ draft pages in the Site Editor, so we may need to work out a design for this in a subsequent issue (perhaps as a part of https://github.com/WordPress/gutenberg/issues/50386).

Another question is which templates to include in the selection modal. `Page` plus any `$custom` templates are a given, as are any patterns associated with the page template. Should we also include _any_ singular template (e.g. `single-post`)?

**Comment by jasmussen** [comment]
That design nicely captures the initial pieces of an i2, but we perhaps need to zero in on the i1, which could be invoked in this way:

<img width="1608" alt="Ways to add" src="https://user-images.githubusercontent.com/1204802/236821512-7a0b6735-23c3-4989-af1b-6b3975ecdb93.png">

To round out the whole flow, that would make for the following 6.3 iteration:

<img width="6784" alt="Adding pages flow" src="https://user-images.githubusercontent.com/1204802/236823467-5b3d38af-9a2d-4585-8d78-fc5ed9f5098d.png">

* Plus to add page
* Slug, parent, draft, edit buttons in modal (same page modal that can be invoked anywhere)
* If you click "Draft", the page is just created, if you click "Edit", you're shown a page template modal

That lets us polish the page detail panel a bit more for later iterations, perhaps letting us include the pattern assembler.

**Comment by mtias** [comment]
Can we try moving the "parent" selector to the empty space of the footer on the left (same line as "edit", "draft"). Also "draft" read weird in this context, can it just be "save"?

**Comment by mtias** [comment]
@jameskoster for the details view iteration, I'd like to explore a design where the available templates are small thumbnails already shown without needing to click on the big template thumbnail.

**Comment by jasmussen** [comment]
> Can we try moving the "parent" selector to the empty space of the footer on the left (same line as "edit", "draft"). Also "draft" read weird in this context, can it just be "save"?

That works well:

<img width="740" alt="Add modal" src="https://user-images.githubusercontent.com/1204802/236853621-aa85617a-34da-4f2f-87d6-b16e145da8d3.png">

<img width="6784" alt="Adding pages flow" src="https://user-images.githubusercontent.com/1204802/236853799-fb4560ca-6f4d-43b8-a833-f24a22117d57.png">

**Comment by mtias** [comment]
Would it be possible to edit the slug or it's just a visualization?

**Comment by jasmussen** [comment]
I was hoping something like this:

<img width="1240" alt="long term" src="https://user-images.githubusercontent.com/1204802/237036022-4bccfe92-f160-49cd-99bb-ed6b656d98e8.png">

But a simpler starting point might be this, click the slug to replace it with an input field for editability:

<img width="1240" alt="near term" src="https://user-images.githubusercontent.com/1204802/237036077-74f7c6cb-d13f-4eec-ba9e-cff30951026c.png">

Main challenge there is to retain a piece of "locked text" — the site — inside that input.

An simpler way to do it would be to just invoke the existing permalink popover from that helper text — this one:

<img width="329" alt="Screenshot 2023-05-09 at 10 15 34" src="https://user-images.githubusercontent.com/1204802/237036548-0976c49e-c63a-4292-baa3-591cc97a6863.png">

**Comment by jasmussen** [comment]
I've updated this issue with mockups for _iteration 1_, and added "Needs dev" label for that.

I've kept the design labels for now, as we still need mockups/feedback for iteration 2.

**Comment by mtias** [comment]
That seems like a whole lot of separate interactions for such a tiny modal. I think if the slug is editable it should probably be a large input. If it's just a reference, then it should not be editable.

**Comment by jasmussen** [comment]
Happy to update. Any preference on editable vs. not?

**Comment by mtias** [comment]
No strong opinion at the moment

**Comment by jasmussen** [comment]
Cool, just visualizing the two options side by side for posterity:

<img width="1240" alt="options" src="https://github.com/WordPress/gutenberg/assets/1204802/1ded2d21-dee7-421c-9ca8-3d28743828f3">

I think we should go with option 1, not editable, for now, and revisit as we go.

**Comment by jameskoster** [comment]
Displaying the slug could increase the likelihood someone would want to edit it. If it's un-editable I wonder if its worth displaying at all?

Does the input need a label?

Does the "Parent" button open a popover? Could be fine, but they can sometimes look a little awkward on top of modals so might be worth mocking up.

"Save" and "Edit" feel a bit unexpected as actions – they're both somewhat ambiguous. How about "Create draft" as the primary action, then show an "Edit" action in the resulting snackbar?

Finally, connecting a dot with https://github.com/WordPress/gutenberg/pull/50444, perhaps "Add" instead of "Create"?

**Comment by SaxonF** [comment]
For iteration one I think we should just show page name and nothing else. We can spend time working on iteration two with more complete controls.

> Does the "Parent" button open a popover? Could be fine, but they can sometimes look a little awkward on top of modals so might be worth mocking up.

More generally I'm not a huge fan of including any controls in the footer of the modal _if its avoidable_ as it throws off the hierarchy of what you are saving/acting on. Related information is fine.

> "Save" and "Edit" feel a bit unexpected as actions – they're both somewhat ambiguous. How about "Create draft" as the primary action, then show an "Edit" action in the resulting snackbar?

Also agree with this. I would just have "Create" as primary action that takes you directly to editor. If we notice people want to bulk create pages we can handle that use case after, perhaps in a similar way to how Linear handles it.

<img width="781" alt="image" src="https://github.com/WordPress/gutenberg/assets/1072756/3136eb32-5a84-4b0a-8fac-53b0e2a66274">

**Comment by richtabor** [comment]
An aside (noting so its shared/visible/I don't loose it): 

Would be neat if there was an option in the "Add a Page" modal to open zoomed out mode + pattern inserter — essentially a "design my own page with patterns" isolated view. Alternatively to starting blank, or picking one page design. 

Related: https://github.com/WordPress/gutenberg/pull/45395

**Comment by SaxonF** [comment]
Kicked off the work [here](https://github.com/WordPress/gutenberg/pull/50565) if someone wants to review. Shouldn't be far off being mergable.

**Comment by mtias** [comment]
I think parent is pretty important to have. Also it should probably say "top level" when it's the default, not "parent" which is confusing.

**Comment by jasmussen** [comment]
I've tentatively updated this one as the conversation seems to have solidified.

**Comment by simison** [comment]
[Issue for categories in the new page template modal](https://github.com/WordPress/gutenberg/issues/46498), neat to see categories in these designs. 👍

**Comment by jameskoster** [comment]
@jasmussen a tiny detail that seems to have been missed: Instead of "Parent", it might be more intuitive if the default option in the dropdown was "Top level". I think that dropdown may need a label too, but it's hard to say for sure without seeing the contents mocked up. I'd echo Saxon's comments about not placing it in the footer, but feel less strongly about that.

The "Save" – and particularly "Edit" – buttons still feel a bit off to me. Could save just be "Create"? If we need the edit option too, maybe "Create & edit"?

**Comment by jasmussen** [comment]
I prefer we start with single words for the buttons, and then change if and when we get to test it in a branch where it might feel off. It's always harder to pull back later on. Same, honestly, with the parent. Also not a strong opinion here, but feels solid enough to start a PR for.

**Comment by getdave** [comment]
Is anyone actively working on the tasks here? The current implementation seem quite tied to the Pages section of the Browse Mode. I envisage that this modal will be needed elsewhere and thus will need to be invoked on a global basis. For example:

- Link UI 
- Navigation in Browse Mode 

I'm thinking the modal needs lifting to somewhere where it can be invoked from an action dispatched to the edit site store.

**Comment by jasmussen** [comment]
I think I've seen some work done by @SaxonF.

**Comment by jameskoster** [comment]
Page creation is now possible via the Pages panel (https://github.com/WordPress/gutenberg/pull/50565). 

Should we close this issue in favor of new ones for any follow-up tasks?

**Comment by SaxonF** [comment]
@getdave the initial implementation of the add page modal was actually global via edit site store like you suggest but was changed to use local state based on feedback. You can view the reversal here https://github.com/WordPress/gutenberg/commit/5fc86cfbc91e0f3963fcda23d9b11370d500c38d . It Might be worth creating a separate issue and could almost copy/paste from earlier commits on that branch, but @youknowriad might have some thoughts on a more elegant global approach .

**Comment by getdave** [comment]
Thanks @SaxonF. Interesting that it was reverted. Was reasoning was provided for that? I assumed it had simply been conceived as a local component from the get-go. 

I can say now that we need to be able to trigger the creation of a new page from both Navigation panel in Browse mode and also the Link Control/Link UI. But in our case we need to be able to control what happens after submission in order that we can act on the resulting data.

**Comment by SaxonF** [comment]
@getdave we just didn't want to introduce deeper changes at that point in time as it was only being triggered in one place. 

[I added a redirectAfterSave](https://github.com/WordPress/gutenberg/pull/50565/commits/f3e097a46bf73ed55bf03e9340fbdf15070c936d) option to the action which you could turn into a more generic onSave callback, but using actions that way feels a little dirty. 

Ideally we'd also want the command center to be able to trigger the add page modal.

**Comment by getdave** [comment]
> @getdave we just didn't want to introduce deeper changes at that point in time as it was only being triggered in one place.

Absolutely. No problem at all. I was just curious if there was another bigger reason I had missed. Thanks for confirming. I commented on the original PR and waiting on Riad to advise if he's happy with this becoming global again.

Thanks for your context and also all the original work here 🙇

**Comment by youknowriad** [comment]
I'm happy if we take the "Add page" modal global (block-editor) again in its own PR. Here are a few things/requirements in my mind:

 - Rely on the existing abstractions (settings) we have in the BlockEditorProvider
 - Use the same UI for all the places we already have (avoid adding a new standard without removing the existing ones)

Related to this, at this point, we have a number of block editor settings and we're often unsure what each setting is about, what format it supports and how is it useful for third-party editors. I think we should also make sure to add a documentation page to document all these settings.

**Comment by getdave** [comment]
> I think we should also make sure to add a documentation page to document all these settings.

@ndiego I wonder if this is something you would be able/interested in looking into?

**Comment by ndiego** [comment]
> @ndiego I wonder if this is something you would be able/interested in looking into?

@getdave definitely. I'll get myself caught up on this issue this week.

**Comment by mtias** [comment]
This looks to be concluded, do we have any follow ups?

**Comment by jameskoster** [comment]
I think just the zoomed-out stuff which is already documented in https://github.com/WordPress/gutenberg/issues/50739.

If anyone disagrees please re-open.

**Comment by ndiego** [comment]
This would be a separate PR, but does it make sense to add the option to choose a specific template for the page on creation?

**Comment by jameskoster** [comment]
I think that is probably covered by https://github.com/WordPress/gutenberg/issues/50739.

**Comment by richtabor** [comment]
Related to https://github.com/WordPress/gutenberg/pull/55117 (connecting dots).
